### PR TITLE
Fix an issue with the last level cache values on Linux running on certain AMD Processors

### LINKED ISSUE #76290: GC picks wrong L3 cache size on Linux
### Description

In https://github.com/dotnet/runtime/issues/48937 it was found that my gen0 budget was 32MiB. Investigating this further, I believe it may even be as high as 64MiB which causes the erratic Gen0 collection times I'm seeing.

System configuration:
- AMD Ryzen 3950x

I wrote a simple app following the implementations that the GC uses on x86 (as far as I can tell) for both Linux and Windows:
- Linux: https://github.com/dotnet/runtime/blob/1955928833e178392f3a40ac1509f0d4a6ca7632/src/coreclr/gc/unix/gcenv.unix.cpp#L880-L899
- Windows: https://github.com/dotnet/runtime/blob/1955928833e178392f3a40ac1509f0d4a6ca7632/src/coreclr/gc/windows/gcenv.windows.cpp#L405-L435

... And found that on Windows it outputs a cache size of 16MiB and on Linux of 64MiB.

I believe the 64MiB value to be incorrectly chosen for this system given the CPU topology:
image

This makes sense, since `_SC_LEVEL3_CACHE_SIZE` returns the total L3 size.

On other platforms, the GC queries `/sys/devices/system/cpu/cpu0/cache/index*/size` to determine the cache size: https://github.com/filipnavara/runtime/blob/1955928833e178392f3a40ac1509f0d4a6ca7632/src/coreclr/gc/unix/gcenv.unix.cpp#L901-L935

... Which results in more reasonable values:
```
$ cat /sys/devices/system/cpu/cpu0/cache/index0/size
32K
$ cat /sys/devices/system/cpu/cpu0/cache/index1/size
32K
$ cat /sys/devices/system/cpu/cpu0/cache/index2/size
512K
$ cat /sys/devices/system/cpu/cpu0/cache/index3/size
16384K
```

### Reproduction Steps

I'm not sure how to extract the Gen0 budget from the GC, so I wrote an app that uses the same method as the GC to determine cache size: https://github.com/smoogipoo/CacheSizeTest

It can be run on Windows and Linux.

Must be run with a multi-CCX CPU such as Ryzen 3950x.

### Expected behavior

The cache size on Linux should be 16MiB.

### Actual behavior

The cache size on Linux is 64MiB.

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by mrsharm**
Fixes: https://github.com/dotnet/runtime/issues/76290

## Problem Details

We recently discovered an issue that affects Unix based VMs where we are taking the host’s (as opposed to the VM’s) for certain AMD processor's last level cache size to be used in the [GetLogicalProcessorCacheSizeFromOS](https://github.com/dotnet/runtime/blob/9b56e737ef9910fcb3b6e80f2bef795459c6982c/src/coreclr/gc/unix/gcenv.unix.cpp#L865) call to discern the gen0 budget for both WKS and SVR. This is because `sysconf`, the method we first try in `GetLogicalProcessorCacheSizeFromOS`, gets us the last level cache of the host machine as opposed to the fallback code path that reads the value of `/sys/devices/system/cpu/cpu0/cache/index{LastLevelCache}/size`. Consequently, the gen0 budgets are significantly different between Unix VMs and Windows using certain AMD processors on the same machine and the further implication of this is that we are probably setting much larger value than expected for the Gen0 budget for the GC running on Unix based VMs. 
 
The details from my v16 CPU based DevBox with an AMD EPYC 7763 64 Core Processor running Ubuntu 22.04.3 via WSL are as follows:
 
- `sysconf` returns the last cache size (L3) for the host machine ([AMD EPYC™ 7763](https://www.amd.com/en/products/processors/server/epyc/7003-series/amd-epyc-7763.html) – specs are here) as 256 MB.
    - Can be repro’d on the command line using:
        `getconf -a | grep “LEVEL3_CACHE_SIZE”` => `LEVEL3_CACHE_SIZE                  268435456`
- Reading `/sys/devices/system/cpu/cpu0/cache/index3/size` returns 32 MiB, the same as the result from GetLogicalProcessorCacheSizeFromOS from Windows that calls [GetLogicalProcessorInformation function (sysinfoapi.h) - Win32 apps | Microsoft Learn](https://learn.microsoft.com/en-us/windows/win32/api/sysinfoapi/nf-sysinfoapi-getlogicalprocessorinformation).
     - Can be repro’d on the command line using lscpu  => `L3:                    32 MiB (1 instance)`

## How To Check for the Issue

1. Get sysconf output: `getconf -a | grep "LEVEL"`
2. The full output of `lscpu`
3. Check if the L3 (or if available, L4) cache size is the same from sysconf and that from lscpu. 
4. If the values are different, the issue exists.

## Solution

- By default, with no configuration changes, first try to read in the cache information from sysfs and if that fails, fall back to the heuristic we use to compute the value for the ARM* cases.
- A new configuration `DOTNET_GCCacheSizeFromSysConf` can be set to 1 to revert to the current behavior i.e., using sysconf to obtain the last level cache.
 
## Performance Testing

Ran with the following GCPerfSim configurations for SVR: `` -tc 28 -tagb 100 -tlgb 0 -lohar 0-pohar 0 -sohsr 100-4000 -lohsr 102400-204800 -pohsr 100-204800 -sohsi 0 -lohsi 0 -pohsi 0 -sohpi 0 -lohpi 0 -sohfi 0 -lohfi 0 -pohfi 0 -allocType reference -testKind time`` 

image

| Metric | Not With SysConf  | With SysConf |
| ------  | --------------------- | -------------- |
| Peak Heap Size (MB) |  339.458 | 2656.194 |
| % Pause Time in GC |  39.0 |  7.7 |

### CODE REVIEW

File: src/coreclr/gc/unix/gcenv.unix.cpp [hunk 1]
```diff
@@ -1042,13 +1079,15 @@ size_t GCToOSInterface::GetCacheSizePerLogicalCpu(bool trueSize)
...
+    bool useSysConf = GCConfig::GetGCCacheSizeFromSysConf();
 
     size_t size = trueSize ? s_maxTrueSize : s_maxSize;
     if (size != 0)
         return size;
 
     size_t maxSize, maxTrueSize;
-    maxSize = maxTrueSize = GetLogicalProcessorCacheSizeFromOS(); // Returns the size of the highest level processor cache
+    maxSize = maxTrueSize = GetLogicalProcessorCacheSizeFromOS(useSysConf); // Returns the size of the highest level processor cache
+    printf ("Last Level Cache Size = %zu\n", maxSize);
```
Comment (mrsharm) [inline src/coreclr/gc/unix/gcenv.unix.cpp:1090]: Will remove this line - using for testing currently.

File: src/coreclr/gc/unix/gcenv.unix.cpp [hunk 2]
```diff
@@ -881,47 +883,107 @@ static size_t GetLogicalProcessorCacheSizeFromOS()
...
+static void GetLogicalProcessorCacheSizeFromSysFs(size_t* cacheLevel, size_t* cacheSize)
+{
+    assert (cacheLevel != nullptr);
+    assert (cacheSize != nullptr);
 
 #if defined(TARGET_LINUX) && !defined(HOST_ARM) && !defined(HOST_X86)
-    if (cacheSize == 0)
+    //
+    // Fallback to retrieve cachesize via /sys/.. if sysconf was not available
+    // for the platform. Currently musl and arm64 should be only cases to use
```
Comment (Maoni0) [inline src/coreclr/gc/unix/gcenv.unix.cpp:902]: this comment needs to be updated to reflect what you are doing.

File: src/coreclr/gc/unix/gcenv.unix.cpp [hunk 3]
```diff
@@ -881,47 +883,107 @@ static size_t GetLogicalProcessorCacheSizeFromOS()
...
+}
+
+static void GetLogicalProcessorCacheSizeFromHeuristic(size_t* cacheLevel, size_t* cacheSize)
+{
+    assert (cacheLevel != nullptr);
+    assert (cacheSize != nullptr);
+
+#if (defined(TARGET_LINUX) && !defined(TARGET_APPLE))
+    {
+        // We expect to get the L3 cache size for Arm64 but currently expected to be missing that info
```
Comment (Maoni0) [inline src/coreclr/gc/unix/gcenv.unix.cpp:939]: this comment needs to be updated to reflect what you are doing.

File: src/coreclr/gc/unix/gcenv.unix.cpp [hunk 4]
```diff
@@ -1042,13 +1079,14 @@ size_t GCToOSInterface::GetCacheSizePerLogicalCpu(bool trueSize)
...
     static volatile size_t s_maxTrueSize;
+    bool useSysConf = GCConfig::GetGCCacheSizeFromSysConf();
 
     size_t size = trueSize ? s_maxTrueSize : s_maxSize;
     if (size != 0)
         return size;
 
     size_t maxSize, maxTrueSize;
-    maxSize = maxTrueSize = GetLogicalProcessorCacheSizeFromOS(); // Returns the size of the highest level processor cache
+    maxSize = maxTrueSize = GetLogicalProcessorCacheSizeFromOS(useSysConf); // Returns the size of the highest level processor cache
```
Comment (Maoni0) [inline src/coreclr/gc/unix/gcenv.unix.cpp:1089]: do we really need to change the signature of `GetLogicalProcessorCacheSizeFromOS`? `GetCacheSizePerLogicalCpu` doesn't have any use of knowing this config value.

**Comment by Maoni0** [comment]
the rest looks okay to me.. would be great if @janvorli could take a look.

**Comment by janvorli** [comment]
There is one thing I keep thinking about. Would it be a problem in case the /sys/devices/system/cpu/cpu0/cache is not present to still read the size from sysconf? In other words, to let the new config knob control just the order in which we try to use the /sys/devices/system/cpu/cpu0/cache and sysconf? I wonder if in the case the /sys/devices/system/cpu/cpu0/cache is missing, the heuristic fallback would give us a reasonable value.

**Comment by Maoni0** [comment]
the way I look at this is from the user's POV the values from sysconf is simply incorrect. so it'd be better to just take the heuristic values. another option is to treat sysconf to always give us the full cache sizes and check to see how many cores this process is actually allowed to use and get a ratio (so if sysconf reports 128mb and 64 cores, our process can only use 8 cores, we take 1/8 of 128mb).

**Comment by janvorli** [comment]
I would rather avoid introducing this new kind of heuristic, because they depend on the internal topology of the processor. One of the cases we were seeing this issue occurred on some AMD processors, because they have 4 separate 3rd level caches where each one is shared by 1/4 of the cores. In this case, the customer was using the full CPU and still the sysconf was returning a sum of the 3rd level cache sizes, it means 4 times higher value. 
So let's keep this change as is.

**Review by Maoni0 — APPROVED** [review]

**Comment by mrsharm** [comment]
## How To Check for the Issue

1. Get sysconf output for the last level cache size: `getconf -a | grep "LEVEL"`
2. The full output of `lscpu`
3. Check if the last level cache size is the same from sysconf and that from lscpu. 
4. If the values are different, the issue exists.

**Pull request merged**